server/channels/app/authorization.go

```diff
@@ -73,26 +73,18 @@ func (a *App) SessionHasPermissionToTeams(c request.CTX, session model.Session,
 	}
 
 	// Getting the list of unique roles from all teams.
-	var roles []string
-	uniqueRoles := make(map[string]bool)
 	for _, teamID := range teamIDs {
 		tm := session.GetTeamByTeamId(teamID)
 		if tm != nil {
-			for _, role := range tm.GetRoles() {
-				uniqueRoles[role] = true
+			if a.RolesGrantPermission(tm.GetRoles(), permission.Id) {
+				continue
 			}
 		}
+		if !a.RolesGrantPermission(session.GetUserRoles(), permission.Id) {
+			return false
+		}
 	}
-
-	for role := range uniqueRoles {
-		roles = append(roles, role)
-	}
-
-	if a.RolesGrantPermission(roles, permission.Id) {
-		return true
-	}
-
-	return a.RolesGrantPermission(session.GetUserRoles(), permission.Id)
+	return true
```

Comment:
I had a very hard time grasping this logic. Given how important this is, I feel this can be simplified a bit and commented heavily.

The complicacy stems from the fact that `a.RolesGrantPermission(session.GetUserRoles(), permission.Id)` needs to be computed just once, yet it is in the for loop because we need to check for the session permissions if a user is not a team member.

Suggestion:
```go
hasSessionPerm := a.RolesGrantPermission(session.GetUserRoles(), permission.Id)
for _, teamID := range teamIDs {
	tm := session.GetTeamByTeamId(teamID)
	if tm != nil {
		// If a team member has permission, then no need to check further.
		if a.RolesGrantPermission(tm.GetRoles(), permission.Id) {
			continue
		}
	}
	// If user is not a team member, or doesn't have permission,
	// check session permission.
	if !hasSessionPerm {
		return false
	}
}
return true
```

cc @streamer45 for a second look.

Reply:
Building on your suggestion, basically, I moved the `RolesGrantPermission` check to the beginning. If that passes, there is no need to check each team.